kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -201,7 +201,8 @@ internal class MutexImpl(locked: Boolean) : Mutex, SelectClause2<Any?, Mutex> {
                         // try lock
                         val update = if (owner == null) EMPTY_LOCKED else Empty(owner)
                         if (_state.compareAndSet(state, update)) { // locked
-                            cont.resume(Unit)
+                            // TODO implement functional type in LockCont as soon as we get rid of legacy JS
+                            cont.resume(Unit) { unlock(owner) }
```

Comment:
I love Kotlin's last param is lambda logic generally. I think in this API, this callsite is an example of where it hides the intent.

```
cont.resume(Unit, onCancellation = { unlock(owner) })
```